File: core/scala/src/main/scala/org/aiddl/core/scala/function/higher_order/MapFunction.scala
package org.aiddl.core.scala.function.higher_order

import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.function.{Evaluator, Function, LazyFunction, DefaultFunctionUri as D}
import org.aiddl.core.scala.representation.{ListTerm, SetTerm, Term, Tuple}

/** Map a function over a collection to create a new collection of function application results.
 *
 * @param c container used for evaluation
 */
protected[function] class MapFunction(eval: Evaluator) extends Function with LazyFunction {
  /** Apply function to each element in a collection term
   *
   * @param x expected to be a tuple composed of a term that evaluates to
   *          a function reference and a collection term to be mapped over
   * @return collection of results of application of function to input collection
   */
  override def apply(x: Term): Term = x match {
    case Tuple(ft, colTerm) => {
      val f = eval(ft).asFunRef
      val collection = eval(colTerm)
      collection match {
        case ListTerm(l) =>
          ListTerm(l.map(f(_)))
        case SetTerm(s) =>
          SetTerm(s.map(f(_)))
        case _ =>
          throw new IllegalArgumentException(s"Cannot map over non collection term $x")
      }
    }
    case _ => throw new IllegalArgumentException(
      s"Bad argument: $x. Need tuple (f c) where f evaluates to " +
        s"a function reference and c is a collection term.")
  }
}

File: core/java/src/main/java/org/aiddl/core/java/function/higher_order/MapFunction.java
package org.aiddl.core.java.function.higher_order;

import java.util.Map;

import org.aiddl.core.java.eval.Evaluator;
import org.aiddl.core.java.function.FunctionRegistry;
import org.aiddl.core.java.function.Uri;
import org.aiddl.core.java.interfaces.ConfigurableFunction;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.interfaces.LazyFunction;
import org.aiddl.core.java.tools.LockableCollection;
import org.aiddl.core.java.tools.LockableList;
import org.aiddl.core.java.tools.LockableSet;
import org.aiddl.core.java.representation.CollectionTerm;
import org.aiddl.core.java.representation.FunctionReferenceTerm;
import org.aiddl.core.java.representation.ListTerm;
import org.aiddl.core.java.representation.Term;

/**
 * Evaluate (map {... } c) to c with all xi substituted by vi.
 * 
 * @author Uwe Koeckemann
 */
public class MapFunction implements LazyFunction, ConfigurableFunction {
	
	Function eval;
	
	/**
	 * Create new evaluator.
	 * @param main evaluator for sub-expression
	 */
	public MapFunction( Evaluator main ) {
		this.eval = main;
	}
	
	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		Term evalName = settings.getOrDefault(Term.sym("eval"), Uri.EVAL);
		this.eval = fReg.getFunction(evalName);
	}
	
	@Override
	public Term apply(Term x) {
		FunctionReferenceTerm fun = this.eval.apply(x.get(0)).asFunRef();
		Function f = fun.getFunction();

		CollectionTerm C = this.eval.apply(x.get(1)).asCollection();
		
		LockableCollection C_m;
		if ( C instanceof ListTerm ) {
			C_m = new LockableList();
		} else {
			C_m = new LockableSet();
		}
		
		for ( Term e : C ) {
			C_m.add(f.apply(e));
		}
		
		if ( C instanceof ListTerm ) {
			return Term.list(C_m);
		} else {
			return Term.set(C_m);
		}
	}
}
